mlflow/system_metrics/system_metrics_monitor.py

```diff
@@ -0,0 +1,130 @@
+
+
+class SystemMetricsMonitor:
+    """Class for monitoring system stats.
+
+    This class is used for pulling system metrics and logging them to MLflow. Calling `start()` will
+    spawn a thread that logs system metrics periodically. Calling `finish()` will stop the thread.
+    Logging is done on a different frequency from pulling metrics, so that the metrics are
+    aggregated over the period. Users can change the logging frequency by setting
+    `$MLFLOW_SYSTEM_METRICS_SAMPLING_INTERVAL` and `$MLFLOW_SYSTEM_METRICS_SAMPLES_BEFORE_LOGGING`
+    environment variables.
+
+    Args:
+        run_id: string, the MLflow run ID.
+        sampling_interval: float, default to 10. The interval (in seconds) at which to pull system
+            metrics.
+        samples_before_logging: int, default to 1. The number of samples to aggregate before
+            logging.
+    """
+
+    def __init__(self, run_id, sampling_interval=10, samples_before_logging=1):
+        from mlflow.utils.autologging_utils import BatchMetricsLogger
+
+        # Instantiate default monitors.
+        self.monitors = [CPUMonitor()]
+        gpu_monitor = GPUMonitor()
+        if gpu_monitor:
+            self.monitors.append(gpu_monitor)
+        self.sampling_interval = MLFLOW_SYSTEM_METRICS_SAMPLING_INTERVAL.get() or sampling_interval
+        self.samples_before_logging = (
+            MLFLOW_SYSTEM_METRICS_SAMPLES_BEFORE_LOGGING.get() or samples_before_logging
+        )
+
+        if isinstance(self.sampling_interval, str):
+            self.sampling_interval = float(self.sampling_interval)
+        if isinstance(self.samples_before_logging, str):
+            self.samples_before_logging = int(self.samples_before_logging)
+
+        self._run_id = run_id
+        self.mlflow_logger = BatchMetricsLogger(self._run_id)
+        self._shutdown_event = threading.Event()
+        self._process = None
+
+    def start(self):
+        """Start monitoring system metrics."""
+        try:
+            self._process = threading.Thread(
+                target=self.monitor,
+                daemon=True,
+                name="SystemMetricsMonitor",
+            )
+            self._process.start()
+            _logger.info("MLflow: started monitoring system metrics.")
+        except Exception as e:
+            _logger.warning(f"MLflow: failed to start monitoring system metrics: {e}")
+            self._process = None
+
+    def monitor(self):
+        """Main monitoring loop, which consistently collect and log system metrics."""
+        from mlflow.tracking.fluent import get_run
+
+        while not self._shutdown_event.is_set():
+            for _ in range(self.samples_before_logging):
+                self.collect_metrics()
+                self._shutdown_event.wait(self.sampling_interval)
+                try:
+                    # Get the MLflow run to check if the run is not RUNNING.
+                    run = get_run(self._run_id)
```

Comment:
Personally I don't love watching the run status in the daemon to kill itself, but note that if you do keep this here, it is somewhat at odds with a comment above:

"The biggest advantage of putting include_system_metrics in the fluent api start_run() over mlflow.system_metrics.start_monitor(run_id) is users don't need to remember to call mlflow.system_metrics.end_monitor(run_id) to kill the logging thread."

Reply:
if the thread kills itself on status, that advantage is eliminated, and can simplify

Reply:
Good catch on the comment, I should remove it.

MLflow run is not guaranteed to exit successfully, it could crash or fail, in which scenario `finish` is not called, so there could be a hanging thread. That's why I am adding this code to detect the associated run's status.